airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/source/relationaldb/state/CursorStateMessageProducer.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.integrations.source.relationaldb.state;
+
+import io.airbyte.cdk.db.IncrementalUtils;
+import io.airbyte.cdk.integrations.source.relationaldb.CursorInfo;
+import io.airbyte.protocol.models.JsonSchemaPrimitiveUtil.JsonSchemaPrimitive;
+import io.airbyte.protocol.models.v0.AirbyteMessage;
+import io.airbyte.protocol.models.v0.AirbyteStateMessage;
+import io.airbyte.protocol.models.v0.AirbyteStreamNameNamespacePair;
+import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream;
+import java.util.Objects;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CursorStateMessageProducer implements SourceStateMessageProducer<AirbyteMessage> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CursorStateMessageProducer.class);
+
+  private final StateManager stateManager;
+  private final String initialCursor;
+  private String currentMaxCursor;
+  private long currentMaxCursorRecordCount = 0L;
+
+  // We keep this field just to control logging frequency.
+  private int totalRecordCount = 0;
+  private AirbyteStateMessage intermediateStateMessage = null;
+
+  private boolean cursorOutOrOrderDetected = false;
+
+  public CursorStateMessageProducer(final StateManager stateManager,
+                                    final String initialCursor) {
+    this.stateManager = stateManager;
+    this.initialCursor = initialCursor;
+    this.currentMaxCursor = initialCursor;
+  }
+
+  /**
+   * @return
+   */
+  @Override
+  public AirbyteStateMessage generateStateMessageAtCheckpoint(final ConfiguredAirbyteStream stream) {
+    // At this stage intermediate state message should never be null; otherwise it would have been
+    // blocked by shouldEmitStateMessage check.
+    final AirbyteStateMessage message = intermediateStateMessage;
+    intermediateStateMessage = null;
+    if (cursorOutOrOrderDetected) {
+      LOGGER.warn("Intermediate state emission feature requires records to be processed in order according to the cursor value. Otherwise, "
+          + "data loss can occur.");
```

Comment:
Can you reword this error message please? As a user, I would be confused. Intermediate state emission isn't a feature that I can opt in or out of. Consider "Out-of-order cursor detected, data loss may occur.". Should this be ERROR instead of WARN?